pgxn/neon/libpagestore.c

```diff
@@ -554,6 +602,15 @@ pg_init_libpagestore(void)
 							   0,	/* no flags required */
 							   check_neon_id, NULL, NULL);
 
+	DefineCustomIntVariable("neon.stripe_size",
```

Comment:
Can we use units of pages here rather than MB?  That way we don't have to remember to do any conversions, and we can safely use a u32 even if we want to use really huge >4GB stripes.  (in the rust code, ShardStripSize  is also in pages)

Reply:
Yes, it is possible too use GUC_UNIT_BLOCKS.
But do we really wan to express stripe in blocks?
Such units are very rarely used in Postgresfor GUCs, just because it is more convenient for users to specify memory size in "standard" units (Kb,Mb).

The same is true for stripe. Even in Rust code:
```
/// Default stripe size in pages: 256MiB divided by 8kiB page size.
const DEFAULT_STRIPE_SIZE: ShardStripeSize = ShardStripeSize(256 * 1024 / 8);
```
You actually define stripe size in Mb and then recalculate tin blocks.

So if you insists - I will change unit type for this GUC. But as far as main criteria for choosing stripe size is really size in bytes hold bone shard rather than number of blocks. Just imagine that decided to change BLCKSZ from 8Kb to 4Kb (it is actually possible in Postgres!). Will it mean that size of stripe should also decreased two times? Most likely not.

Reply:
This isn't a big deal, but I think we'd better keep life simple by using a page count across all the components -- that way we can avoid debates about units, and avoid risk of possible mistakes in conversions when going between components.

>You actually define stripe size in Mb and then recalculate it in blocks.

That's just a comment to help readers get an intuition of how big a stripe is.  Everywhere in the code, it's in pages: there are no conversions -- avoiding conversions between units is beneficial,  because it avoids a possible source of bugs.

> it is more convenient for users to specify memory size in "standard" units (Kb,Mb).

Makes sense for regular postgres features, but this is an internal value that will always be set by a machine: in our environment this config is always going to be set by the control plane.  If we ever exposed it to a user, that would be way up in the Console UI somewhere, where we could decide in the UI how we wanted to format/convert the value.

> Just imagine that decided to change BLCKSZ from 8Kb to 4Kb (it is actually possible in Postgres!). Will it mean that size of stripe should also decreased two times? Most likely not.

Interesting to think about -- we'd probably do some testing to establish our new stripe size: it's not obvious to me that we'd want to keep the byte size constant, because some overheads are per-page, when doing things like compaction in the pageserver.

Reply:
> Such units are very rarely used in Postgresfor GUCs, just because it is more convenient for users to specify memory size in "standard" units (Kb,Mb).

There are a lot of GUCs that use GUC_UNIT_BLOCKS. Like `shared_buffers`. The user can still set them in kilobytes or megabytes, e.g. `123 MB` or `1 GB`, and the system will report them in kB/MB/GB as appropriate.

Reply:
I'm surprised that stripe_size is a GUC. Does it ever change, could it be just a compiled in constant? Or do we want to experiment with different stripe sizes? Surely you can't change it on the fly, so PGC_SIGHUP is even more surprising.

Is there a concept of a primary pageserver? The one where you get the basebackup? Could we get the stripe size from the primary pageserver at startup, so that you wouldn't need to set it correctly as a GUC? Perhaps include the stripe size somewhere in the basebackup.

Reply:
> I'm surprised that stripe_size is a GUC. 

It was suggestion by @jcsp 
From my point of view compile time constant will be enough.

Reply:
> There are a lot of GUCs that use GUC_UNIT_BLOCKS. Like shared_buffers. The user can still set them in kilobytes or megabytes, e.g. 123 MB or 1 GB, and the system will report them in kB/MB/GB as appropriate.

Ok, I can change unit of this GUC to blocks. 
But my point is that stripe size doesn't somehow depends on number of blocks. 
The main criteria is size in bytes. And so it seems to be quote storage to specify stripe size blocks. 
For example `wal_segment_size` is specified in bytes, not blocks, although `XLOG_BLCKSZ = BLCKSZ`

Reply:
> Ok, I can change unit of this GUC to blocks.

Looks like this didn't happen yet, please make the change in a followup PR to avoid blocking this one.

Reply:
It looks like you've updated this, closing thread